### ISSUE TITLE
D3D8 draw call framebuffer inconsistency

### ISSUE THREAD

**Issue by AnyOldName3** [apitrace/apitrace#676]
While investigating an old D3D8 application, Morrowind, I realised that *sometimes* when inspecting the state of a draw call, it would show the framebuffer before the draw call, and other times it would show the framebuffer after the draw call. I'm *pretty* sure that with OpenGL it always shows the state after the call.

This was pretty irritating as it mislead me about how the application was rendering a particular dumb effect.

**Comment by jrfonseca** [comment]
This shouldn't happen, and I don't see how it can happen.

d3dretrace always dumps the state after call where the state is being requested, except on present calls, where it tries to dump state before the present call (as presenting can often cause the contents to be discarded after the present.)

So if one inspects the state on a draw call, it should be the state after the draw.  If one inspects the state on a call before the draw, it should be the state before the draw.

**Comment by AnyOldName3** [comment]
If I thought this was how it was supposed to work, I wouldn't have opened the issue report, but it was what was happening.

One thing that might affect Morrowind, but wouldn't affect most applications, was that I was *probably* using MGE XE when recording the trace. MGE XE is an engine hack that hooks various functions with the aim of adding new graphical features. One of the things I think it does is use some D3D9 features despite the original engine using D3D8.1, so it's possible that however it does that triggers an edge case either when recording or replaying a trace.

**Comment by jrfonseca** [comment]
I'll need a concrete trace and specific call number where the issue can be seen then, to investigate it further.

There was a problem like this for multithreaded apps, but it was fixed on 80e37c7fcc5e3e867fa7ec0dea48bc0c7ecfeb1d long time ago. 

Regarding MGE XE, I don't know how it works, my guess is that it overrides d3d8.dll or some other game DLL.  But unless it's intercepting apitrace's d3dretrace.exe it shouldn't cause wrong state from being dumped.

**Comment by AnyOldName3** [comment]
As it was nearly a year ago that I observed this, it might take some time to track things down. I'll see what I can do, though.

**Comment by AnyOldName3** [comment]
I actually still had a trace created on the same day as I reported this, so I just had a look at it. Thankfully it was pretty obvious which draw calls I'd been interested in, but I tried the equivalent call in four different frames and couldn't reproduce this. There are another 400 frames with an equivalent call that I haven't tested, so it's possible I just got unlucky when I ran into this and picked a bad frame by chance, or it's possible something else has changed in the meantime that's fixed it unintentionally.

Just in case someone wants to try every single frame for me (I don't imagine it would be *too* difficult to get d3dretrace to dump as JSON and then diff each call, then search those diffs for a non-draw call immediately following a draw call where the second call is the one that changes the framebuffer) I'm going to link the trace.

https://mega.nz/file/LwYRwASB#jxQFsDw_6yRw4lvJPHlDWo6dmMxpCy5ksX9_MsOunyo

The calls I'd have been looking at are the nearer sphere, either the first pass, which draws it with a soil texture, e.g. 626202, or the second pass that applies some grass as a decal, e.g. 626207.

Having written all this, though, I see that the commit you linked above was only three weeks old when I reported this, and it's pretty likely the binaries I was using were at least three weeks old. I think that's a pretty strong sign that this was fixed and I should have tried reproing in a newer build before I reported this.
